src/bootstrap/src/ferrocene/run.rs

```diff
@@ -99,3 +100,139 @@ impl Step for TraceabilityMatrix {
+        let _cargo_dir = builder.cargo_dir();
+
+        if builder.config.dry_run() {
+            return;
+        }
+
+        let core_tests_binary_path = get_test_binary_path(builder).unwrap();
+
+        if coverage_report_data_dir.exists() {
+            let mut files = coverage_report_data_dir.read_dir().expect("Failed to read dir");
+
+            let has_profraw_files = files.any(|file| {
+                if let Ok(file) = file {
+                    file.path().extension().map(|ext| ext.to_str().unwrap()) == Some("profraw")
+                } else {
+                    false
+                }
+            });
+
+            if !has_profraw_files {
+                panic!("No profraw files found in build/coverage directory");
+            }
+        }
+
+        if coverage_report_out_dir.exists() {
+            builder.remove_dir(&coverage_report_out_dir);
+            builder.info(&format!("Removed previous report at {:?}", coverage_report_out_dir));
+        }
+        let rustc_path = builder.out.join(builder.config.build.triple).join("stage1/bin/rustc");
+
+        let grcov_installed = Command::new("grcov").arg("--version").output().is_ok();
+
+        if !grcov_installed {
+            panic!("grcov not installed, to install it run:\n cargo install grcov");
+        }
+
+        let mut cmd = builder.tool_cmd(Tool::FerroceneGenerateCoverageReport);
+        cmd.env("COVERAGE_REPORT_DATA_DIR", coverage_report_data_dir);
+        cmd.env("COVERAGE_REPORT_OUT_DIR", coverage_report_out_dir);
+        cmd.env("COVERAGE_REPORT_SRC_PATH", coverage_src_path);
+        cmd.env("COVERAGE_REPORT_BIN_PATH", core_tests_binary_path);
+        cmd.env("RUSTC", rustc_path);
+
+        builder.run(&mut cmd);
+    }
+
+    fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
+        run.path("ferrocene/tools/generate-coverage-report")
+    }
+
+    fn make_run(run: RunConfig<'_>) {
+        run.builder.ensure(GenerateCoverageReport);
+    }
+}
+fn env_llvm_profile_data_dir() -> Option<PathBuf> {
+    // If the LLVM_PROFILE_FILE environment variable is set, get the directory
+    match std::env::var("LLVM_PROFILE_FILE") {
+        Ok(path) => {
+            let file_path = Path::new(&path);
+            let directory = file_path.parent().unwrap_or(Path::new("./"));
+            Some(directory.to_owned())
+        }
+        Err(_err) => None,
+    }
+}
+fn get_test_binary_path(builder: &Builder<'_>) -> Option<PathBuf> {
+    let cargo_toml_str =
+        std::fs::read_to_string("library/core/Cargo.toml").expect("Failed to read Cargo.toml");
+    let cargo_toml_data: toml::Value =
+        toml::from_str(&cargo_toml_str).expect("Failed to parse Cargo.toml");
+
+    let tests = cargo_toml_data.get("test").expect("Failed to get tests");
+    let tests_array = tests.as_array();
+    let tests_array = tests_array.unwrap();
+    let test = &tests_array[0];
+    let test = test.as_table().unwrap();
```

Comment:
Can this be fetched from `builder.crates`? The source code that populates it is in https://github.com/ferrocene/ferrocene/blob/main/src/bootstrap/src/core/metadata.rs.

Reply:
The `crates` field is populated in https://github.com/ferrocene/ferrocene/blob/main/src/bootstrap/src/core/metadata.rs. That files parses the output of `cargo metadata` ([fields reference](https://doc.rust-lang.org/nightly/cargo/commands/cargo-metadata.html)) and stores it in the field.

If the information you need is present in `cargo metadata` you can add the relevant serde structs/fields in the file to parse it! Looking at the fields reference, it should provide all the information needed for all the test binaries it runs.

Reply:
Since we have only one test binary name for now, can we just hardcode the binary name as "coretests" ?
Given that we are using the code coverage for core library only at this point in time.

Reply:
Yeah that's fine.